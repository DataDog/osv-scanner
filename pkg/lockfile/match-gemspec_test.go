package lockfile_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/google/osv-scanner/internal/testutility"

	"github.com/google/osv-scanner/pkg/lockfile"
	"github.com/google/osv-scanner/pkg/models"
	"github.com/stretchr/testify/assert"
)

var gemspecFileMatcher = lockfile.GemspecFileMatcher{}

func TestGemspecFileMatcher_GetSourceFile_FileDoesNotExist(t *testing.T) {
	t.Parallel()

	lockFile, err := lockfile.OpenLocalDepFile("fixtures/bundler/no-gemspec/Gemfile.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := gemspecFileMatcher.GetSourceFile(lockFile)
	assert.Nil(t, sourceFile)
	assert.NoError(t, err)
}

func TestGemspecFileMatcher_GetSourceFile(t *testing.T) {
	t.Parallel()
	dir, err := os.Getwd()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	basePath := "fixtures/bundler/gemspec/"
	sourcefilePath := filepath.FromSlash(filepath.Join(dir, basePath+"test.gemspec"))

	lockFile, err := lockfile.OpenLocalDepFile(basePath + "Gemfile.lock")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := gemspecFileMatcher.GetSourceFile(lockFile)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	assert.Equal(t, sourcefilePath, sourceFile.Path())
}

func TestGemspecFileMatcher_Match(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/bundler/gemspec/test.gemspec")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "base64",
			Version:        "0.2.0",
			PackageManager: models.Bundler,
		},
		{
			Name:           "json",
			Version:        "2.9.1",
			PackageManager: models.Bundler,
		},
		{
			Name:           "thor",
			Version:        "1.3.2",
			PackageManager: models.Bundler,
		},
		{
			Name:           "timeout",
			Version:        "0.4.3",
			PackageManager: models.Bundler,
		},
		{
			Name:           "useragent",
			Version:        "0.16.11",
			PackageManager: models.Bundler,
		},
		{
			Name:           "websocket-driver",
			Version:        "0.7.7",
			PackageManager: models.Bundler,
		},
		{
			Name:           "websocket-extensions",
			Version:        "0.1.5",
			PackageManager: models.Bundler,
		},
		{
			Name:           "zeitwerk",
			Version:        "2.7.1",
			PackageManager: models.Bundler,
		},
	}

	err = gemspecFileMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().WithJSONNormalization().MatchJSON(t, packages)
}
