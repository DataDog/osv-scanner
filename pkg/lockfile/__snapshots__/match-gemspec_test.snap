
[TestGemspecFileMatcher_Match - 1]
[
  {
    "name": "base64",
    "version": "0.2.0",
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "packageManager": "Bundler"
  },
  {
    "name": "json",
    "version": "2.9.1",
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "packageManager": "Bundler"
  },
  {
    "name": "thor",
    "version": "1.3.2",
    "blockLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 3,
        "end": 38
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "versionLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 31,
        "end": 38
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 23,
        "end": 23
      },
      "column": {
        "start": 23,
        "end": 29
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "timeout",
    "version": "0.4.3",
    "blockLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 3,
        "end": 49
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "versionLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 42,
        "end": 49
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 24,
        "end": 24
      },
      "column": {
        "start": 31,
        "end": 40
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "useragent",
    "version": "0.16.11",
    "blockLocation": {
      "line": {
        "start": 25,
        "end": 27
      },
      "column": {
        "start": 3,
        "end": 42
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "versionLocation": {
      "line": {
        "start": 25,
        "end": 27
      },
      "column": {
        "start": 44,
        "end": 42
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 25,
        "end": 25
      },
      "column": {
        "start": 31,
        "end": 42
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "websocket-driver",
    "version": "0.7.7",
    "depGroups": [
      "dev"
    ],
    "blockLocation": {
      "line": {
        "start": 36,
        "end": 36
      },
      "column": {
        "start": 3,
        "end": 53
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 36,
        "end": 36
      },
      "column": {
        "start": 35,
        "end": 53
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "websocket-extensions",
    "version": "0.1.5",
    "depGroups": [
      "dev"
    ],
    "blockLocation": {
      "line": {
        "start": 30,
        "end": 35
      },
      "column": {
        "start": 3,
        "end": 4
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "versionLocation": {
      "line": {
        "start": 30,
        "end": 35
      },
      "column": {
        "start": 59,
        "end": 4
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 30,
        "end": 30
      },
      "column": {
        "start": 35,
        "end": 57
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  },
  {
    "name": "zeitwerk",
    "version": "2.7.1",
    "depGroups": [
      "dev"
    ],
    "blockLocation": {
      "line": {
        "start": 29,
        "end": 29
      },
      "column": {
        "start": 3,
        "end": 67
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "versionLocation": {
      "line": {
        "start": 29,
        "end": 29
      },
      "column": {
        "start": 47,
        "end": 67
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "nameLocation": {
      "line": {
        "start": 29,
        "end": 29
      },
      "column": {
        "start": 35,
        "end": 45
      },
      "file_name": "<rootdir>/fixtures/bundler/gemspec/test.gemspec"
    },
    "packageManager": "Bundler"
  }
]
---
